File: python/laplace.py
from solver import laplace
from mesh import Mesh
import numpy as np
import time
import os

# %% Functions

def meshparam(elem,size,xyMax):
    
    # Node coordinates
    
    eId = []
    node = size+1
    nXY = np.zeros((node**2,2))
    idx = [i for i in range(size,node**2-node-1,node)]
    for i in range(node): nXY[i*node:(i+1)*node] = [[j,i] for j in range(node)]
    nXY *= xyMax/size
    
    # Element indices
    
    for i in range(node**2-node-1):
        if (i not in idx):
            if (elem==4): eId += [[i,i+1,i+node+1,i+node]]
            if (elem==3): eId += [[i,i+1,i+node]]+[[i+1,i+node+1,i+node]]

    # Node indices for BC
    
    n1 = np.arange(node)
    n2 = np.arange(node,node**2,node)
    n3 = (np.arange(node)+size*node)[1:]
    n4 = (np.arange(node,node**2,node)-1)[1:]
    nId = np.sort(np.concatenate((n1,n2,n3,n4)))
    return nXY,eId,nId
        
# %% Mesh And Flux

elem = 3
xyMax = 1
size = 50

fun = lambda x,y: np.sin(2*x)+np.sin(2*y)
nXY,eId,nId = meshparam(elem,size,xyMax)

data = {}
data['bc'] = 0
data['nId'] = nId

# %% Solver

start = time.time()
mesh = Mesh(nXY,eId,fun)
print('Mesh:',time.time()-start,'[sec]')

start = time.time()
u = laplace(mesh,data)
print('Solver:',time.time()-start,'[sec]')

# %% Writes Solution

path = r'../output' 
if not os.path.exists(path): os.makedirs(path)
np.save('../output/solution.npy',u)
np.save('../output/nXY.npy',nXY)

File: code/laplace.cpp
#include <Eigen/Dense>
#include <iostream>
#include <direct.h>
#include "solver.h"
#include <fstream>
#include <vector>
#include "mesh.h"

using namespace std;
using namespace Eigen;

struct Param{

    vector<vector<double>> nXY;
    vector<vector<int>> eId;
    vector<int> nId;
};

Param meshParam(int elem,int size,double xyMax){

    int k = 0;
    Param param;
    vector<int> idx;
    int node = size+1;
    vector<vector<int>> eId;
    vector<int> nId(4*size);
    vector<vector<double>> nXY(node*node,vector<double>(2));
    for(int i=size; i<node*size-1; i+=node){idx.push_back(i);}

    // Node coordinates

    for(int i=0; i<node; i++){
        for(int j=0; j<node; j++){
            nXY[i*node+j] = {j*xyMax/size,i*xyMax/size};
        }
    }

    // Element indices

    for(int i=0; i<node*size-1; i++){if(i!=idx[k]){

        if(elem==4){eId.push_back({i,i+1,i+node+1,i+node});}
        if(elem==3){eId.push_back({i,i+1,i+node});eId.push_back({i+1,i+node+1,i+node});}}
        else{k++;}
    }

    // Boundary node indices

    for(int i=0; i<size; i++){
        
        nId[i] = i;
        nId[size+i] = (i+1)*node-1;
        nId[2*size+i] = node*node-i-1;
        nId[3*size+i] = (size-i)*node;
    }

    param.nXY = nXY;
    param.eId = eId;
    param.nId = nId;
    return param;
}

// Solves the laplace equation Δu(x,y) = f(x,y)

int main(){

    int type = 3;
    int size = 50;
    double xyMax = 1;

    Param param = meshParam(type,size,xyMax);
    auto fun = [](MatrixXd xy){return (2*xy.col(0)).array().sin()+(2*xy.col(1)).array().sin();};
    vector<double> bc(param.nId.size(),0);
    
    Data data;
    data.fun = fun;
    data.bcDir = bc;
    data.nId = param.nId;

    // Mesh and solver

    Mesh mesh(param.nXY,param.eId);
    cout << "\nMesh: done" << endl;
    vector<VectorXd> u = laplace(mesh,data);
    cout << "Solver: done" << endl;

    // Writes the file

    mkdir("../output");
    ofstream solution("../output/solution.txt");
    ofstream nXY("../output/nXY.txt");

    for (int i=0; i<param.nXY.size(); i++){
        nXY << param.nXY[i][0] << "," << param.nXY[i][1] << "\n";
    }

    for (int i=0; i<u.size(); i++){
        for (int j=0; j<u[0].size()-1; j++){solution << u[i][j] << ",";}
        solution << u[i][u[0].size()-1] << "\n";
    }

    cout << "Writing: done\n" << endl;
    return 0;
}